### PR TITLE
[Core feature] Allow dictionaries to be passed to execute workflows that take dataclasses

### PR DESCRIPTION
**Pull request by jasonlai1218**
## Tracking issue
https://github.com/flyteorg/flyte/issues/4098

## Describe your changes
- Import the `json` module
- Add a check for dictionary types in the `DataclassTransformer` class
- Raise a `TypeTransformerFailedError` if a key in the dictionary does not match any field in the dataclass
- Convert the dictionary to a JSON string using `json.dumps()`

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Setup Process

1. write a `data_class2.py`
```
import typing
from dataclasses import dataclass
from dataclasses_json import dataclass_json
from flytekit import workflow, task


@dataclass_json
@dataclass
class Data:
    a: int = 10
    b: str = "default_b"
    c: typing.Optional[str] = "default_c"


@task
def concat(x: list[Data]):
    for data in x:
        print(f"{data.a}_{data.b}_{data.c}")


@workflow
def my_wf(x: list[Data]) -> int:
    concat(x=x)
    return 0
```

2. register data_class2.py
execute cmd: `pyflyte register -p flytesnacks --domain development --version 4 data_class2.py`

3. write a `remote_wf2.py`
```
from flytekit.remote import FlyteRemote
from flytekit.configuration import Config


if __name__ == "__main__":
    # FlyteRemote object is the main entrypoint to API
    remote = FlyteRemote(
        config=Config.for_endpoint(endpoint="localhost:30080", insecure=True),
        default_project="flytesnacks",
        default_domain="development",
    )

    # Get Task
    flyte_workflow = remote.fetch_workflow(name="data_class2.my_wf", version="4")

    # Run Task
    execution = remote.execute(
        flyte_workflow,
        inputs={'x': [{"a": 10, "b": "hello"}, {"a": 100, "b": "hello2"}]}
    )

```
and execute the program

### Successful screenshot

Screenshot 2023-12-01 at 1 23 55 PM

---

Screenshot 2023-12-01 at 1 24 02 PM

---

### FlyteFile

1. write a `flytefile_data_class.py`

```
import csv
import os
from collections import defaultdict
from dataclasses import dataclass
from typing import List

from dataclasses_json import dataclass_json

import flytekit
from flytekit import task, workflow
from flytekit.types.file import FlyteFile


@dataclass_json
@dataclass
class Data:
    csv_url: FlyteFile
    columns_to_normalize: List[str]
    column_names: List[str]
    output_location: str


@task
def normalize_columns(data: Data) -> FlyteFile:
    # read the data from the raw csv file
    parsed_data = defaultdict(list)
    with open(data.csv_url, newline="\n") as input_file:
        reader = csv.DictReader(input_file, fieldnames=data.column_names)
        next(reader)  # Skip header
        for row in reader:
            for column in data.columns_to_normalize:
                parsed_data[column].append(float(row[column].strip()))

    # normalize the data
    normalized_data = defaultdict(list)
    for colname, values in parsed_data.items():
        mean = sum(values) / len(values)
        std = (sum([(x - mean) ** 2 for x in values]) / len(values)) ** 0.5
        normalized_data[colname] = [(x - mean) / std for x in values]

    # write to local path
    out_path = os.path.join(
        flytekit.current_context().working_directory,
        f"normalized-{os.path.basename(data.csv_url.path).rsplit('.')[0]}.csv",
    )
    with open(out_path, mode="w") as output_file:
        writer = csv.DictWriter(output_file, fieldnames=data.columns_to_normalize)
        writer.writeheader()
        for row in zip(*normalized_data.values()):
            writer.writerow({k: row[i] for i, k in enumerate(data.columns_to_normalize)})

    if data.output_location:
        return FlyteFile(path=out_path, remote_path=data.output_location)
    else:
        return FlyteFile(path=out_path)


@workflow
def normalize_csv_file(
    data: Data,
) -> FlyteFile:
    return normalize_columns(data=data)
```

2. register flytefile_data_class.py
`pyflyte register -p flytesnacks --domain development --version 1 flytefile_data_class.py`

3. write a `flytefile_remote_wf.py`
```
from flytekit.remote import FlyteRemote
from flytekit.configuration import Config

if __name__ == "__main__":
    # FlyteRemote object is the main entrypoint to API
    remote = FlyteRemote(
        config=Config.for_endpoint(endpoint="localhost:30080", insecure=True),
        default_project="flytesnacks",
        default_domain="development",
    )

    # Get Task
    flyte_workflow = remote.fetch_workflow(name="flytefile_data_class.normalize_csv_file", version="1")

    # Run Task
    execution = remote.execute(
        flyte_workflow,
        inputs={
            'data': {
                "csv_url": {"path": "https://people.sc.fsu.edu/~jburkardt/data/csv/biostats.csv"},
                "column_names": ["Name", "Sex", "Age", "Heights (in)", "Weight (lbs)"],
                "columns_to_normalize": ["Age"],
                "output_location": ""
            },
        }
    )
```
and execute the program

### Successful screenshot
Screenshot 2023-12-08 at 6 11 42 PM
Screenshot 2023-12-08 at 6 11 51 PM

---

### StructuredDataset

1. write a `structureddataset_data_class.py`

```
from dataclasses_json import dataclass_json
from flytekit.types.structured import StructuredDataset

from flytekit import task, workflow, kwtypes
import pandas as pd


@dataclass_json
@dataclass
class Data:
    structured_dataset: StructuredDataset


col = kwtypes(Age=int)


@task
def get_subset_pandas_df(data: Data) -> StructuredDataset:
    df = data.structured_dataset.open(pd.DataFrame).all()
    df = pd.concat([df, pd.DataFrame([[30]], columns=["Age"])])
    return StructuredDataset(dataframe=df)


@task
def print_df(df: pd.DataFrame):
    print(df)


@task
def save_df(df: pd.DataFrame, path: str):
    df.to_csv(path, index=False)


@workflow
def simple_sd_wf(data: Data) -> StructuredDataset:
    # pandas_df = generate_pandas_df(a=19)
    output = get_subset_pandas_df(data=data)
    print_df(df=output.dataframe)
    save_df(df=output.dataframe, path="test3.csv")
    return output.dataframe
```

2. register flytefile_data_class.py
`pyflyte register -p flytesnacks --domain development --version 1 structureddataset_data_class.py`

3. upload parquet file to minio
Screenshot 2023-12-08 at 6 26 49 PM

4. write a `structureddataset_remote_wf.py`
```
from flytekit.remote import FlyteRemote
from flytekit.configuration import Config


if __name__ == "__main__":
    # FlyteRemote object is the main entrypoint to API
    remote = FlyteRemote(
        config=Config.for_endpoint(endpoint="localhost:30080", insecure=True),
        default_project="flytesnacks",
        default_domain="development",
    )

    # Get Task
    flyte_workflow = remote.fetch_workflow(name="structureddataset_data_class.simple_sd_wf", version="1")

    # Run Task
    execution = remote.execute(
        flyte_workflow,
        inputs={
            'data': {
                "structured_dataset": {
                    "uri": "s3://test/part-00000-8cbb2a62-33cc-475a-8895-b06e75aa1db6-c000.snappy.parquet",
                    "file_format": "parquet",
                }
            },
        }
    )
```
and execute the program

### Successful screenshot
Screenshot 2023-12-08 at 6 24 39 PM
Screenshot 2023-12-08 at 6 25 00 PM

---

### Scenario 1: Parameter type error
Screenshot 2023-12-01 at 1 20 52 AM

---

Screenshot 2023-12-01 at 12 21 58 PM

```
error message:
TypeTransformerFailedError: Type of Val '<class 'int'>' is not an instance of 
<class 'str'>
```

---

### Scenario 2: Wrong field name
Screenshot 2023-12-01 at 12 23 55 PM

```
error message:
TypeTransformerFailedError: The original fields are missing the following keys 
from the dataclass fields: ['b']
```

---

Screenshot 2023-12-01 at 12 25 02 PM

```
TypeTransformerFailedError: The original fields are missing the following keys 
from the dataclass fields: ['b']
```

---

Screenshot 2023-12-01 at 12 25 57 PM

```
TypeTransformerFailedError: The original fields have the following extra keys 
that are not in dataclass fields: ['d']
```

### CODE REVIEW

File: tests/flytekit/unit/core/test_type_engine.py [hunk 1]
```diff
@@ -1404,6 +1405,65 @@ class Bar(DataClassJsonMixin):
...
+    vd = {"x": 3, "y": "hello"}
+    DataclassTransformer().assert_type(Args, vd)
+
+    # Test when v is a dict but missing keys from dataclass
+    md = {"x": 3}
+    with pytest.raises(
+        TypeTransformerFailedError,
+        match=re.escape("The original fields are missing the following keys from the dataclass fields: ['y']"),
+    ):
+        DataclassTransformer().assert_type(Args, md)
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_type_engine.py:1424]: `y` is optional, so we don't necessary need to specify it, right?

File: tests/flytekit/unit/core/test_type_engine.py [hunk 2]
```diff
@@ -1404,6 +1405,65 @@ class Bar(DataClassJsonMixin):
         DataclassTransformer().assert_type(gt, pv)
 
 
+def test_assert_dict_type():
+    @dataclass
+    class Args(DataClassJsonMixin):
+        x: int
+        y: typing.Optional[str]
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_type_engine.py:1412]: 1. if `y` is optional, we should give it a default value.
2. could we also add `flytefile`, `StructuredDataset` and `dataclass` to this class?
```suggestion
        y: typing.Optional[str] = None
        f: FlyteFile
        s: StructuredDataset
        d: AnotherDataclass
```

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -424,21 +456,24 @@ def get_literal_type(self, t: Type[T]) -> LiteralType:
...
-            type(python_val), DataClassJSONMixin
-        ):
-            raise TypeTransformerFailedError(
-                f"Dataclass {python_type} should be decorated with @dataclass_json or inherit DataClassJSONMixin to be "
-                f"serialized correctly"
-            )
-        self._serialize_flyte_type(python_val, python_type)
+        if isinstance(python_val, dict):
+            json_str = json.dumps(python_val)
+        else:
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:461]: nit:
```suggestion
        if isinstance(python_val, dict):
            json_str = json.dumps(python_val)
            return Literal(scalar=Scalar(generic=_json_format.Parse(json_str, _struct.Struct())))
```

File: tests/flytekit/unit/core/test_type_engine.py (hunk 1 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_type_engine.py:1424]: Yes, you're right.

File: tests/flytekit/unit/core/test_type_engine.py (hunk 2 above)
Comment (jasonlai1218) [reply tests/flytekit/unit/core/test_type_engine.py:1416]: Thanks for the suggestion, I've added it.

File: flytekit/core/type_engine.py [hunk 4]
```diff
@@ -349,20 +350,61 @@ def assert_type(self, expected_type: Type[DataClassJsonMixin], v: T):
...
+            optional_keys = {k for k, t in expected_fields_dict.items() if UnionTransformer.is_optional_type(t)}
 
-            val = v.__getattribute__(f.name)
-            if dataclasses.is_dataclass(val):
-                self.assert_type(expected_type, val)
-            elif original_type != expected_type:
-                raise TypeTransformerFailedError(f"Type of Val '{original_type}' is not an instance of {expected_type}")
+            # Remove the Optional keys from the keys of original_dict
+            keys1 = set(original_dict.keys()) - optional_keys
+            keys2 = set(expected_fields_dict.keys()) - optional_keys
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:361]: nit: could we name `key1` and `key2` to `original_key` and `expected_key`?

File: flytekit/core/type_engine.py [hunk 5]
```diff
@@ -349,20 +350,61 @@ def assert_type(self, expected_type: Type[DataClassJsonMixin], v: T):
...
+
+            # Check if dict1 is missing any keys from dict2
```
Comment (pingsutw) [inline flytekit/core/type_engine.py:363]:
```suggestion
            # Check if original_key is missing any keys from expected_key
```

File: flytekit/core/type_engine.py (hunk 4 above)
Comment (jasonlai1218) [reply flytekit/core/type_engine.py]: updated

File: flytekit/core/type_engine.py (hunk 5 above)
Comment (jasonlai1218) [reply flytekit/core/type_engine.py]: updated

**Review by pingsutw — DISMISSED** [review]
Thanks @jasonlai1218. Could you resolve the merge conflict?

**Comment by jasonlai1218** [comment]
updated

**Review by pingsutw — DISMISSED** [review]

**Review by pingsutw — APPROVED** [review]

**Pull request merged**